go-controller/pkg/ovn/gateway.go

```diff
@@ -49,18 +53,42 @@ func (ovn *Controller) createGatewayVIPs(protocol kapi.Protocol, sourcePort int3
 			klog.Errorf("Gateway router %s does not have physical ip (%v)", gatewayRouter, err)
 			continue
 		}
+
+		vips := physicalIPs
+		if len(svcIPs) > 0 {
+			vips = svcIPs
+		}
+		// If self ip is in target list, we need to use special IP to allow hairpin back to host
+		newTargets := util.UpdateIPsSlice(targetIPs, physicalIPs, []string{types.V4HostMasqueradeIP, types.V6HostMasqueradeIP})
+
 		// With the physical_ip:sourcePort as the VIP, add an entry in
 		// 'load_balancer'.
-		err = ovn.createLoadBalancerVIPs(loadBalancer, physicalIPs, sourcePort, targetIPs, targetPort)
+		err = ovn.createLoadBalancerVIPs(gatewayLB, vips, sourcePort, newTargets, targetPort)
 		if err != nil {
-			klog.Errorf("Failed to create VIP in load balancer %s - %v", loadBalancer, err)
+			klog.Errorf("Failed to create VIP in load balancer %s - %v", gatewayLB, err)
 			continue
 		}
+
+		if config.Gateway.Mode == config.GatewayModeShared {
```

Comment:
these things always bit us later :(

Reply:
it's a necessary evil for now. Will go away once we deprecate local gw mode and remove it.